test/app-luatest/uri_test.lua

```diff
@@ -21,39 +21,10 @@ local build_ascii_table = function()
     return ascii_tbl, table.concat(ascii_tbl, "")
 end
 
-local function str_to_map(str)
-    local tbl = {}
-    str:gsub(".", function(c)
-        tbl[c] = true
-    end)
-    return tbl
-end
-
--- This is a *dumb* implementation of function for escaping a single symbol
--- used as a test oracle.
-local function expected_escaped(str, opts)
-    assert(type(str) == "string" and #str == 1)
-    if opts.plus and str == " " then
-        return "+"
-    elseif str_to_map(opts.unreserved)[str] then
-        return str
-    else
-        return string.format("%%%02X", string.byte(str))
-    end
-end
-
-g_opts.test_uri_escape = function(cg)
-    local ascii_tbl, _ = build_ascii_table()
-    local opts = uri[cg.params.name]
-    for _, ch in pairs(ascii_tbl) do
-        t.assert_equals(uri.escape(ch, opts), expected_escaped(ch, opts))
-    end
-end
```

Comment:
The test leans on internal options representation and I guess it is the reason why it was deleted with changing the implementation. Without the test we can miss some corner cases. For example, a typo in building a symbol table may lead to a problem with 255th symbol. I also think that it would be nice to freeze another corner case: '\x00` encoding.

I propose to hold the testing data in the test.

```lua
local escaped_bytes = {
    RFC3986 = table.concat({
        [[%00%01%02%03%04%05%06%07%08%09%0A%0B%0C%0D%0E%0F]],
        [[%10%11%12%13%14%15%16%17%18%19%1A%1B%1C%1D%1E%1F]],
        [[%20%21%22%23%24%25%26%27%28%29%2A%2B%2C-.%2F]],
        [[0123456789%3A%3B%3C%3D%3E%3F]],
        [[%40ABCDEFGHIJKLMNO]],
        [[PQRSTUVWXYZ%5B%5C%5D%5E_]],
        [[%60abcdefghijklmno]],
        [[pqrstuvwxyz%7B%7C%7D~%7F]],
        [[%80%81%82%83%84%85%86%87%88%89%8A%8B%8C%8D%8E%8F]],
        [[%90%91%92%93%94%95%96%97%98%99%9A%9B%9C%9D%9E%9F]],
        [[%A0%A1%A2%A3%A4%A5%A6%A7%A8%A9%AA%AB%AC%AD%AE%AF]],
        [[%B0%B1%B2%B3%B4%B5%B6%B7%B8%B9%BA%BB%BC%BD%BE%BF]],
        [[%C0%C1%C2%C3%C4%C5%C6%C7%C8%C9%CA%CB%CC%CD%CE%CF]],
        [[%D0%D1%D2%D3%D4%D5%D6%D7%D8%D9%DA%DB%DC%DD%DE%DF]],
        [[%E0%E1%E2%E3%E4%E5%E6%E7%E8%E9%EA%EB%EC%ED%EE%EF]],
        [[%F0%F1%F2%F3%F4%F5%F6%F7%F8%F9%FA%FB%FC%FD%FE%FF]]}, ''),
    PATH = table.concat({
        [[%00%01%02%03%04%05%06%07%08%09%0A%0B%0C%0D%0E%0F]],
        [[%10%11%12%13%14%15%16%17%18%19%1A%1B%1C%1D%1E%1F]],
        [[%20!%22%23$%25&'()*+,-./]],
        [[0123456789:;%3C=%3E%3F]],
        [[@ABCDEFGHIJKLMNO]],
        [[PQRSTUVWXYZ%5B%5C%5D%5E_]],
        [[%60abcdefghijklmno]],
        [[pqrstuvwxyz%7B%7C%7D~%7F]],
        [[%80%81%82%83%84%85%86%87%88%89%8A%8B%8C%8D%8E%8F]],
        [[%90%91%92%93%94%95%96%97%98%99%9A%9B%9C%9D%9E%9F]],
        [[%A0%A1%A2%A3%A4%A5%A6%A7%A8%A9%AA%AB%AC%AD%AE%AF]],
        [[%B0%B1%B2%B3%B4%B5%B6%B7%B8%B9%BA%BB%BC%BD%BE%BF]],
        [[%C0%C1%C2%C3%C4%C5%C6%C7%C8%C9%CA%CB%CC%CD%CE%CF]],
        [[%D0%D1%D2%D3%D4%D5%D6%D7%D8%D9%DA%DB%DC%DD%DE%DF]],
        [[%E0%E1%E2%E3%E4%E5%E6%E7%E8%E9%EA%EB%EC%ED%EE%EF]],
        [[%F0%F1%F2%F3%F4%F5%F6%F7%F8%F9%FA%FB%FC%FD%FE%FF]]}, ''),
    PATH_PART = table.concat({
        [[%00%01%02%03%04%05%06%07%08%09%0A%0B%0C%0D%0E%0F]],
        [[%10%11%12%13%14%15%16%17%18%19%1A%1B%1C%1D%1E%1F]],
        [[%20!%22%23$%25&'()*+,-.%2F]],
        [[0123456789:;%3C=%3E%3F]],
        [[@ABCDEFGHIJKLMNO]],
        [[PQRSTUVWXYZ%5B%5C%5D%5E_]],
        [[%60abcdefghijklmno]],
        [[pqrstuvwxyz%7B%7C%7D~%7F]],
        [[%80%81%82%83%84%85%86%87%88%89%8A%8B%8C%8D%8E%8F]],
        [[%90%91%92%93%94%95%96%97%98%99%9A%9B%9C%9D%9E%9F]],
        [[%A0%A1%A2%A3%A4%A5%A6%A7%A8%A9%AA%AB%AC%AD%AE%AF]],
        [[%B0%B1%B2%B3%B4%B5%B6%B7%B8%B9%BA%BB%BC%BD%BE%BF]],
        [[%C0%C1%C2%C3%C4%C5%C6%C7%C8%C9%CA%CB%CC%CD%CE%CF]],
        [[%D0%D1%D2%D3%D4%D5%D6%D7%D8%D9%DA%DB%DC%DD%DE%DF]],
        [[%E0%E1%E2%E3%E4%E5%E6%E7%E8%E9%EA%EB%EC%ED%EE%EF]],
        [[%F0%F1%F2%F3%F4%F5%F6%F7%F8%F9%FA%FB%FC%FD%FE%FF]]}, ''),
    QUERY = table.concat({
        [[%00%01%02%03%04%05%06%07%08%09%0A%0B%0C%0D%0E%0F]],
        [[%10%11%12%13%14%15%16%17%18%19%1A%1B%1C%1D%1E%1F]],
        [[%20!%22%23$%25&'()*+,-./]],
        [[0123456789%3A;%3C=%3E?]],
        [[%40ABCDEFGHIJKLMNO]],
        [[PQRSTUVWXYZ%5B%5C%5D%5E_]],
        [[%60abcdefghijklmno]],
        [[pqrstuvwxyz%7B%7C%7D~%7F]],
        [[%80%81%82%83%84%85%86%87%88%89%8A%8B%8C%8D%8E%8F]],
        [[%90%91%92%93%94%95%96%97%98%99%9A%9B%9C%9D%9E%9F]],
        [[%A0%A1%A2%A3%A4%A5%A6%A7%A8%A9%AA%AB%AC%AD%AE%AF]],
        [[%B0%B1%B2%B3%B4%B5%B6%B7%B8%B9%BA%BB%BC%BD%BE%BF]],
        [[%C0%C1%C2%C3%C4%C5%C6%C7%C8%C9%CA%CB%CC%CD%CE%CF]],
        [[%D0%D1%D2%D3%D4%D5%D6%D7%D8%D9%DA%DB%DC%DD%DE%DF]],
        [[%E0%E1%E2%E3%E4%E5%E6%E7%E8%E9%EA%EB%EC%ED%EE%EF]],
        [[%F0%F1%F2%F3%F4%F5%F6%F7%F8%F9%FA%FB%FC%FD%FE%FF]]}, ''),
    QUERY_PART = table.concat({
        [[%00%01%02%03%04%05%06%07%08%09%0A%0B%0C%0D%0E%0F]],
        [[%10%11%12%13%14%15%16%17%18%19%1A%1B%1C%1D%1E%1F]],
        [[%20!%22%23$%25%26'()*+,-./]],
        [[0123456789%3A;%3C%3D%3E?]],
        [[%40ABCDEFGHIJKLMNO]],
        [[PQRSTUVWXYZ%5B%5C%5D%5E_]],
        [[%60abcdefghijklmno]],
        [[pqrstuvwxyz%7B%7C%7D~%7F]],
        [[%80%81%82%83%84%85%86%87%88%89%8A%8B%8C%8D%8E%8F]],
        [[%90%91%92%93%94%95%96%97%98%99%9A%9B%9C%9D%9E%9F]],
        [[%A0%A1%A2%A3%A4%A5%A6%A7%A8%A9%AA%AB%AC%AD%AE%AF]],
        [[%B0%B1%B2%B3%B4%B5%B6%B7%B8%B9%BA%BB%BC%BD%BE%BF]],
        [[%C0%C1%C2%C3%C4%C5%C6%C7%C8%C9%CA%CB%CC%CD%CE%CF]],
        [[%D0%D1%D2%D3%D4%D5%D6%D7%D8%D9%DA%DB%DC%DD%DE%DF]],
        [[%E0%E1%E2%E3%E4%E5%E6%E7%E8%E9%EA%EB%EC%ED%EE%EF]],
        [[%F0%F1%F2%F3%F4%F5%F6%F7%F8%F9%FA%FB%FC%FD%FE%FF]]}, ''),
    FRAGMENT = table.concat({
        [[%00%01%02%03%04%05%06%07%08%09%0A%0B%0C%0D%0E%0F]],
        [[%10%11%12%13%14%15%16%17%18%19%1A%1B%1C%1D%1E%1F]],
        [[%20!%22%23$%25&'()*+,-./]],
        [[0123456789%3A;%3C=%3E?]],
        [[%40ABCDEFGHIJKLMNO]],
        [[PQRSTUVWXYZ%5B%5C%5D%5E_]],
        [[%60abcdefghijklmno]],
        [[pqrstuvwxyz%7B%7C%7D~%7F]],
        [[%80%81%82%83%84%85%86%87%88%89%8A%8B%8C%8D%8E%8F]],
        [[%90%91%92%93%94%95%96%97%98%99%9A%9B%9C%9D%9E%9F]],
        [[%A0%A1%A2%A3%A4%A5%A6%A7%A8%A9%AA%AB%AC%AD%AE%AF]],
        [[%B0%B1%B2%B3%B4%B5%B6%B7%B8%B9%BA%BB%BC%BD%BE%BF]],
        [[%C0%C1%C2%C3%C4%C5%C6%C7%C8%C9%CA%CB%CC%CD%CE%CF]],
        [[%D0%D1%D2%D3%D4%D5%D6%D7%D8%D9%DA%DB%DC%DD%DE%DF]],
        [[%E0%E1%E2%E3%E4%E5%E6%E7%E8%E9%EA%EB%EC%ED%EE%EF]],
        [[%F0%F1%F2%F3%F4%F5%F6%F7%F8%F9%FA%FB%FC%FD%FE%FF]]}, ''),
    FORM_URLENCODED = table.concat({
        [[%00%01%02%03%04%05%06%07%08%09%0A%0B%0C%0D%0E%0F]],
        [[%10%11%12%13%14%15%16%17%18%19%1A%1B%1C%1D%1E%1F]],
        [[+%21%22%23%24%25%26%27%28%29*%2B%2C-.%2F]],
        [[0123456789%3A%3B%3C%3D%3E%3F]],
        [[%40ABCDEFGHIJKLMNO]],
        [[PQRSTUVWXYZ%5B%5C%5D%5E_]],
        [[%60abcdefghijklmno]],
        [[pqrstuvwxyz%7B%7C%7D%7E%7F]],
        [[%80%81%82%83%84%85%86%87%88%89%8A%8B%8C%8D%8E%8F]],
        [[%90%91%92%93%94%95%96%97%98%99%9A%9B%9C%9D%9E%9F]],
        [[%A0%A1%A2%A3%A4%A5%A6%A7%A8%A9%AA%AB%AC%AD%AE%AF]],
        [[%B0%B1%B2%B3%B4%B5%B6%B7%B8%B9%BA%BB%BC%BD%BE%BF]],
        [[%C0%C1%C2%C3%C4%C5%C6%C7%C8%C9%CA%CB%CC%CD%CE%CF]],
        [[%D0%D1%D2%D3%D4%D5%D6%D7%D8%D9%DA%DB%DC%DD%DE%DF]],
        [[%E0%E1%E2%E3%E4%E5%E6%E7%E8%E9%EA%EB%EC%ED%EE%EF]],
        [[%F0%F1%F2%F3%F4%F5%F6%F7%F8%F9%FA%FB%FC%FD%FE%FF]]}, ''),
}
```

It is generated by the following one-liner and a bit post-processed to look better:

```lua
tarantool> do local what = FORM_URLENCODED local plus = what.plus local unreserved = what.unreserved local function find(haystack, needle) for i = 1, #haystack do if haystack:sub(i, i) == needle then return true end end return false end local string_buffer = {} for i = 0, 255 do local c = string.char(i) if plus and c == ' ' then string_buffer[i] = '+' elseif find(unreserved, c) then string_buffer[i] = c else string_buffer[i] = '%' .. c:hex():upper() end end return table.concat(string_buffer, '') end
```

Alternatively we can just copy `unreserved` value for each class and test against it instead of `opts.unreserved`. Or we can use our knowledge about cdata's content and access it. But I think that holding testing data in a test is better (in any form).

Reply:
Nice!